elastic/apm-server
# Move building of binaries, Docker images, and packages to Makefile

## Motivation/summary

Move building of binaries, Docker images, and and packages out of Mage and into our Makefile. The goals are to simplify the build process, take greater ownership of the way in which we build and package APM Server, and better integrate with tools such as Skaffold (which can build a non-templated Dockerfile) for a better local development experience.

We introduce our own Dockerfile. This is based on the beats templated Dockerfile, with the following major changes:

 - We commit the `tini` binaries to the repo, to avoid having to download and verify during every image build step. The binaries are, as expected, tiny.
 - The Dockerfile is not templated, but uses build args where necessary. This would allow us to better integrate with tools like Skaffold, automatically building when source has changed.
 - We build the apm-server binary in Docker. This means the Docker image can be built without any prior steps.
 - We use BuildKit features to cache the Go module and Go build cache directories. On Linux, this requires
   setting DOCKER_BUILDKIT=1; on macOS and Windows this is the default with Docker Desktop.
 - We no longer explicitly create the "apm-server" user, instead we just set the UID to 1000. This simplifies the build, removing the need for installing packages. By avoiding the addition of any packages, we can switch the base image from Ubuntu to RedHat UBI8 without any other changes.

These are the major changes related to building binaries and packages:

 - Mage is no longer used at all, and has been removed as a dependency.
 - Static files added to packages are no longer templated.
 - We now use [nfpm](https://nfpm.goreleaser.com/), rather than [fpm](https://github.com/jordansissel/fpm). nfpm is a pure Go alternative to fpm, which means we have no dependency on Docker for building packages.
 - We now inject the Git commit timestamp into binaries for the "build time", rather than the system time at which the build happened. This means our Go binaries are now reproducible. We continue to use the system time for the Docker image builds: the base images may change from time to time (i.e. tags may be replaced), so they are not fully reproducible anyway.
- Docker image tarballs, and dependencies.csv, are now named according to the expectations of the unified release manager configuration.

Packaging targets have been renamed from "release-manager-release" and "release-manager-snapshot" to "package" and "package-snapshot" respectively. As we now use Make exclusively, building binaries and packages can run in parallel with `make -j<N> package`.

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

## How to test these changes

1. Install and smoke test each package and Docker image.
2. Confirm the commit hash and date are reported in the apm-server logs.
3. Compare IronBank Docker build context tarball to one built from the 8.4 branch, verify there are no discrepancies.
4. Build an apm-server binary multiple times, confirm that the build ID remains the same.
5. Confirm apm-server binaries are stripped
6. Confirm 64-bit apm-server are built with PIE, and 32-bit binaries are not

## Related issues

Closes https://github.com/elastic/apm-server/issues/8644
